File: crates/saju-lib/src/timezone.rs
use chrono::{DateTime, FixedOffset, NaiveDateTime, Offset, TimeZone, Utc};
use chrono_tz::Tz;
use std::str::FromStr;

#[derive(Clone, Copy, Debug)]
pub enum TimeZoneSpec {
    Fixed(FixedOffset),
    Named(Tz),
}

impl TimeZoneSpec {
    pub fn localize(&self, naive: NaiveDateTime) -> Result<DateTime<FixedOffset>, String> {
        use chrono::offset::LocalResult;
        match self {
            TimeZoneSpec::Fixed(offset) => offset
                .from_local_datetime(&naive)
                .single()
                .ok_or_else(|| "invalid local time for fixed offset".to_string()),
            TimeZoneSpec::Named(tz) => match tz.from_local_datetime(&naive) {
                LocalResult::Single(dt) => Ok(dt.with_timezone(&dt.offset().fix())),
                LocalResult::Ambiguous(dt1, _) => Ok(dt1.with_timezone(&dt1.offset().fix())),
                LocalResult::None => Err("local time does not exist in this timezone".to_string()),
            },
        }
    }

    pub fn to_local(&self, utc: DateTime<Utc>) -> DateTime<FixedOffset> {
        match self {
            TimeZoneSpec::Fixed(offset) => utc.with_timezone(offset),
            TimeZoneSpec::Named(tz) => {
                let dt = tz.from_utc_datetime(&utc.naive_utc());
                dt.with_timezone(&dt.offset().fix())
            }
        }
    }

    pub fn name(&self) -> String {
        match self {
            TimeZoneSpec::Fixed(offset) => format!("{}", offset),
            TimeZoneSpec::Named(tz) => tz.name().to_string(),
        }
    }
}

pub fn parse_timezone(input: &str) -> Result<TimeZoneSpec, String> {
    if let Some(offset) = parse_fixed_offset(input) {
        return Ok(TimeZoneSpec::Fixed(offset));
    }
    if let Ok(tz) = Tz::from_str(input) {
        return Ok(TimeZoneSpec::Named(tz));
    }
    Err("timezone must be IANA name (e.g., Asia/Seoul) or offset (+09:00)".to_string())
}

fn parse_fixed_offset(input: &str) -> Option<FixedOffset> {
    let trimmed = input.trim();
    let sign = if trimmed.starts_with('+') {
        1
    } else if trimmed.starts_with('-') {
        -1
    } else {
        return None;
    };
    let rest = &trimmed[1..];
    let (hours, minutes) = if let Some((h, m)) = rest.split_once(':') {
        (h, m)
    } else if rest.len() == 4 {
        (&rest[0..2], &rest[2..4])
    } else {
        return None;
    };
    let hours: i32 = hours.parse().ok()?;
    let minutes: i32 = minutes.parse().ok()?;
    if hours.abs() > 23 || minutes.abs() > 59 {
        return None;
    }
    let total = sign * (hours * 3600 + minutes * 60);
    FixedOffset::east_opt(total)
}

File: packages/saju-lib/src/timezone.ts
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc.js';
import timezone from 'dayjs/plugin/timezone.js';

dayjs.extend(utc);
dayjs.extend(timezone);

export interface TimeZoneSpec {
  type: 'fixed' | 'named';
  offsetSeconds?: number;
  name?: string;
}

export function parseTimezone(input: string): TimeZoneSpec {
  const fixed = parseFixedOffset(input);
  if (fixed !== null) {
    return { type: 'fixed', offsetSeconds: fixed };
  }
  // Try to validate as IANA timezone
  const d = dayjs.tz('2000-01-01 00:00', input);
  if (d.isValid()) {
    return { type: 'named', name: input };
  }
  throw new Error('timezone must be IANA name (e.g., Asia/Seoul) or offset (+09:00)');
}

function parseFixedOffset(input: string): number | null {
  const trimmed = input.trim();
  let sign: number;
  if (trimmed.startsWith('+')) {
    sign = 1;
  } else if (trimmed.startsWith('-')) {
    sign = -1;
  } else {
    return null;
  }
  const rest = trimmed.slice(1);
  let hours: number;
  let minutes: number;
  if (rest.includes(':')) {
    const [h, m] = rest.split(':');
    hours = parseInt(h, 10);
    minutes = parseInt(m, 10);
  } else if (rest.length === 4) {
    hours = parseInt(rest.slice(0, 2), 10);
    minutes = parseInt(rest.slice(2, 4), 10);
  } else {
    return null;
  }
  if (isNaN(hours) || isNaN(minutes) || Math.abs(hours) > 23 || Math.abs(minutes) > 59) {
    return null;
  }
  return sign * (hours * 3600 + minutes * 60);
}

export function tzName(spec: TimeZoneSpec): string {
  if (spec.type === 'named') return spec.name!;
  const total = spec.offsetSeconds!;
  const sign = total >= 0 ? '+' : '-';
  const abs = Math.abs(total);
  const h = Math.floor(abs / 3600);
  const m = Math.floor((abs % 3600) / 60);
  return `${sign}${String(h).padStart(2, '0')}:${String(m).padStart(2, '0')}`;
}

export function localize(spec: TimeZoneSpec, dateStr: string, timeStr: string): dayjs.Dayjs {
  const naive = `${dateStr} ${timeStr}`;
  if (spec.type === 'named') {
    return dayjs.tz(naive, spec.name!);
  }
  return dayjs.utc(naive).subtract(spec.offsetSeconds!, 'second');
}

export function toLocal(spec: TimeZoneSpec, utcDate: dayjs.Dayjs | Date): dayjs.Dayjs {
  const d = utcDate instanceof Date ? dayjs.utc(utcDate) : utcDate;
  if (spec.type === 'named') {
    return d.tz(spec.name!);
  }
  return d.utcOffset(spec.offsetSeconds! / 60);
}

export function getOffsetSeconds(spec: TimeZoneSpec, dateStr: string, timeStr: string): number {
  const dt = localize(spec, dateStr, timeStr);
  return dt.utcOffset() * 60;
}
